mono/sgen/sgen-alloc.c

```diff
@@ -246,6 +250,8 @@ sgen_alloc_obj_nolock (GCVTable vtable, size_t size)
 
 				zero_tlab_if_necessary (p, size);
 			} else {
+				THREAD_BYTES_ALLOCATED += TLAB_NEXT - TLAB_START - size;
```

Comment:
I see you are bumping the size once the previous tlab is full. This does not handle in real time the allocation inside the current tlab. We could do this by computing on the spot the difference between tlab_next and tlab_start (when asked for the alloc size).

Reply:
> This does not handle all allocation paths. You should look for all paths that alloc large objects (sgen_los_alloc_large_inner) and all paths that allocate directly inside the major heap (alloc_degraded, alloc_small_pinned_obj). It also doesn't handle the lock free fast path (sgen_try_alloc_obj_nolock)

So, many of these allocator call other allocator functions. I am actually incrementing on some sgen_los_alloc_large_inner calls, just not inside the method. To avoid double counting, I adopt this convention: Only increment on the paths where we actually allocate, and not if allocation is deferred to another method, to avoid confusion. Except for the hot path, where we increment on retiring the TLAB.

Reply:
Hmm, no, what I have suggested doesn't actually work well. Because, for example, when we allocate on promotion I would double count those if I push increments down into the mark sweep allocator. I'll just have to do it carefully and make sure I get everything with no double counts.